### PR TITLE
Prometheus: Variable query editor improvements

### PR DESCRIPTION
**Pull request by bohandley**
Fixes https://github.com/grafana/grafana/issues/69859

**What is this feature?**
The Prom variable query editor allows a user to choose a `label_values` query type. This allows a user to choose a label name and an optional metric. These are the following updates to the `label_values` query type inputs.

- [x] Use the metric select in the Prom query variable editor.
- [x] Allow the metric to be filtered with optional label filters.
- [x] Allow the metric to be clearable.
- [x] When a metric is selected, filter the options of required label name select dropdown
- [x] Add variables to the options in the select drop downs (labels and metric for the label_values query type). 
- [x] Styling changes and copy

Screenshot 2023-06-11 at 7 03 17 PM

Screenshot 2023-06-11 at 7 03 27 PM

**Why do we need this feature?**
The previous metric input for the label_values query type was a text input that was too small and the metric was not able to be filtered. There was also a recent PR to allow for the label_values label input to accept variables. These changes listed above add functionality to the query variable editor and allow the customer greater flexibility in creating a variable.

**Who is this feature for?**
Anyone using the prom query variable editor.

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by gtk-grafana** [comment]
@bohandley looks like this might address https://github.com/grafana/grafana/issues/69859

**Comment by itsmylife** [comment]
I check out the branch and open my existing dashboard with a bunch of variables in it. 

Manual testing results: 
- When I keep adding new label filters the row gets wider. I expect it to be wrapped. 
- After I select a `Label` I expect to see the results. I have to click another place to trigger to run the query

**Comment by itsmylife** [comment]
I think the definition column will be filled in another PR, right?
image

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+export function MetricsLabelsSection({
+  datasource,
+  query,
+  onChange,
+  onBlur,
+  variableEditor,
+}: MetricsLabelsSectionProps) {
+  // fixing the use of 'as' from refactoring
+  // @ts-ignore
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx:31]: ~~I think the type is `QueryBuilderLabelFilter[]`. Do we need `@ts-ignore` here?~~

I see that it is about `LabelFilter.tsx` and specifically it is about `onLabelsChange` function. Could you please check if that's possible to have the proper types? Let's avoid `ts-ignore` as much as we can.

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+  /**
+   * Function kicked off when user interacts with label in label filters.
+   * Formats a promQL expression and passes that off to helper functions depending on API support
+   * @param forLabel
+   */
+  const onGetLabelNames = async (forLabel: Partial<QueryBuilderLabelFilter>): Promise<SelectableValue[]> => {
+    // If no metric we need to use a different method
+    if (!query.metric) {
+      // Todo add caching but inside language provider!
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx:62]: @gtk-grafana I see it in this PR. Do we track this todo in somewhere?

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx [hunk 3]
```diff
@@ -57,29 +61,32 @@ export function MetricSelect({
     },
   ];
 
-  const customFilterOption = useCallback((option: SelectableValue<any>, searchQuery: string) => {
-    const label = option.label ?? option.value;
-    if (!label) {
-      return false;
-    }
+  const customFilterOption = useCallback(
+    (option: SelectableValue<any>, searchQuery: string) => {
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx:65]: Sorry this is my OCD :) Do we need that indentation?

**Review by itsmylife — APPROVED** [review]
Some nitpicks but the rest is 🚀

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx (hunk 2 above)
Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx:62]: We can remove the todo, the `prometheusResourceBrowserCache` will cache this

File: public/app/plugins/datasource/prometheus/querybuilder/components/PromQueryBuilder.tsx [hunk 4]
```diff
@@ -208,23 +40,7 @@ export const PromQueryBuilder = React.memo<Props>((props) => {
...
-          onGetMetrics={onGetMetrics}
-          datasource={datasource}
-          labelsFilters={query.labels}
-          metricLookupDisabled={datasource.lookupsDisabled}
-        />
-        <LabelFilters
-          debounceDuration={datasource.getDebounceTimeInMilliseconds()}
-          getLabelValuesAutofillSuggestions={getLabelValuesAutocompleteSuggestions}
-          labelsFilters={query.labels}
-          // eslint-ignore
```
Comment (bohandley) [inline public/app/plugins/datasource/prometheus/querybuilder/components/PromQueryBuilder.tsx:223]: Link for discussion about not using ts-ignore

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        labelsFilters={query.labels}
+        metricLookupDisabled={datasource.lookupsDisabled}
+        onBlur={onBlur ? onBlur : () => {}}
+        variableEditor={variableEditor}
+      />
+      <LabelFilters
+        debounceDuration={datasource.getDebounceTimeInMilliseconds()}
+        getLabelValuesAutofillSuggestions={getLabelValuesAutocompleteSuggestions}
+        labelsFilters={query.labels}
+        onChange={onChangeLabels}
```
Comment (bohandley) [inline public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx:217]: Link for discussion about not using ts-ignore

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx (hunk 1 above)
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx:31]: @ itsmylife This is what we had before

```
onChange={onChangeLabels as (labelFilters: Array<Partial<QueryBuilderLabelFilter>>) => void}
```
https://github.com/grafana/grafana/pull/69884/files#r1229794280

and 
```
  const onChangeLabels = (labels: QueryBuilderLabelFilter[]) => {
    onChange({ ...query, labels });
  };
```
https://github.com/grafana/grafana/pull/69884/files#r1229795429

When I moved this into a new file, using `onChangeLabels as` was not permitted by prettier. I removed the `as` and using the type `QueryBuilderLabelFilter[]` this caused a type error. 
https://github.com/grafana/grafana/pull/69884/files#r1229803709

I am not sure how to fix this as the function definition is the same in the LabelFilters.tsx component but prettier won't allow this.

**Comment by bohandley** [comment]
@itsmylife 
> I check out the branch and open my existing dashboard with a bunch of variables in it.
> 
> Manual testing results:
> 
> * When I keep adding new label filters the row gets wider. I expect it to be wrapped.
This is the result of using the InlineField component to add the label. It does not allow wrapping. I could override the `flex-direction: row` to be `flex-direction: column` and we would have this which doesn't look good because of the wide button. 

Screenshot 2023-06-14 at 3 38 14 PM

The EditorField does allow for wrapping but comes with a different style of label. This is inconsistent with the rest of the labels. I would be interested in changing the labels, this would match the rest of the template variable editor style. 
Screenshot 2023-06-14 at 3 41 19 PM

> * After I select a `Label` I expect to see the results. I have to click another place to trigger to run the query
The labels aren't actually part of the query, so I left out the functionality to run the query onChange of labels. They are only there to filter the metrics list and they do filter the metrics on change.

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx (hunk 3 above)
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx:65]: I tried to remove the indentation but prettier added it back. Placing the function argument inline with the useCallback function makes the line too long. Therefore, we get the indent 🤷🏻

**Comment by bohandley** [comment]
@itsmylife wrap is working, looks better now 🎁 

https://github.com/grafana/grafana/assets/25674746/69f976e9-239a-4146-9b74-cb70e8fed6fc

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx (hunk 1 above)
Comment (itsmylife) [reply public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx:31]: I do not have any suggestions but I checked it briefly and realized that it is not a simple change to make. So we can leave it like this and address it later on.

**Comment by itsmylife** [comment]
The wrapping looks good. 

I experienced strange behavior though. Probably not related to your changes but let's see. 
I have a dashboard that has many variables defined. I clicked variables and click one of the existing ones which has the query type "label values". To test wrapping I clicked the plus button multiple times. It works nice and well. But after a while (a couple of seconds later) all my empty label filters are gone. I see that in the network tab, there are some queries that keep running in the background. Dashboard has a 5s refresh setting. Each time I got the result from `api/v1/series` endpoint the filters are gone.

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx (hunk 1 above)
Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx:31]: Can we create an issue with the tech-debt label to address this?

Comment (bohandley) [reply public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx:31]: I will create the issue once this PR is merged to track this.

**Comment by bohandley** [comment]
> The wrapping looks good.
> 
> I experienced strange behavior though. Probably not related to your changes but let's see. I have a dashboard that has many variables defined. I clicked variables and click one of the existing ones which has the query type "label values". To test wrapping I clicked the plus button multiple times. It works nice and well. But after a while (a couple of seconds later) all my empty label filters are gone. I see that in the network tab, there are some queries that keep running in the background. Dashboard has a 5s refresh setting. Each time I got the result from `api/v1/series` endpoint the filters are gone.

I was not able to reproduce this. It could be a timing issue due to dashboard setup? I have tried this with a prometheus instance and cortex but cannot reproduce this. After our discussion on slack, we are attributing this to a dashboard setup. I will make a note of this issue for future work. Thank you!

https://github.com/grafana/grafana/assets/25674746/48e37533-6bbc-4b7a-8cc9-2280bec21b10

**Pull request merged**